### PR TITLE
Add snmp polling capability for Memcached

### PR DESCRIPTION
**Pull request by florianbeer**
This is a sort of proof of concept of snmp's extend feature. It is here to show, that no check_mk is needed to run local scripts on remote servers, it's all available within snmp.  
The apache, nginx and mysql scripts in `/opt/librenms/includes/polling/application` already have snmp polling included, I just ported it to the memcache poller script.

The slightly daunting nested code beginning on line 30 is just a transformation to get memcached's strange status output into the required php array. It is basically a map/reduce for all of the available values using native php functions.

To activate memcache over snmp take the following steps:

- Deploy the memcache status script (https://github.com/librenms/librenms-agent/blob/master/agent-local/memcached) on your remote server for example in `/usr/local/bin`
- On your remote server edit the snmp daemon config (most probably in `/etc/snmp/snmpd.conf`) and add the following line at the bottom `extend memcached /usr/local/memcached`
- Restart snmpd
- On your LibreNMS webfrontend, find the device you just deployed the script to, click on edit and activate "Memcached" on the "Applications" tab.

### CODE REVIEW

File: includes/polling/applications/memcached.inc.php [hunk 1]
```diff
@@ -1,6 +1,53 @@
 <?php
 
-$data = $agent_data['app']['memcached'][$app['app_instance']];
+if (!empty($agent_data['app']['memcached'])) {
+    $data = $agent_data['app']['memcached'][$app['app_instance']];
+} else {
+    $options = '-O qv';
+    $oid     = 'nsExtendOutputFull.9.109.101.109.99.97.99.104.101.100';
+    $res     = explode(';' , snmp_get($device, $oid, $options));
```
Comment (murrant) [inline includes/polling/applications/memcached.inc.php]: Wouldn't it be easier to remove the first line `<<<app-memcached>>>`, and then call `unserialize()` on the remainder of the string?

**Comment by murrant** [comment]
Make sure you update the documentation as well: `doc/Extensions/Applications.md`

File: includes/polling/applications/memcached.inc.php [hunk 2]
```diff
@@ -1,6 +1,53 @@
 <?php
 
-$data = $agent_data['app']['memcached'][$app['app_instance']];
+if (!empty($agent_data['app']['memcached'])) {
+    $data = $agent_data['app']['memcached'][$app['app_instance']];
+} else {
+    $options = '-O qv';
+    $oid     = 'nsExtendOutputFull.9.109.101.109.99.97.99.104.101.100';
```
Comment (murrant) [inline includes/polling/applications/memcached.inc.php]: Use the full number `.1.3.6.1.4.1.8072.1.3.2.3.1.2.9.109.101.109.99.97.99.104.101.100` or specify the `NET-SNMP-EXTEND-MIB` mib.

File: includes/polling/applications/memcached.inc.php (hunk 1 above)
Comment (florianbeer) [inline includes/polling/applications/memcached.inc.php]: Had I known, that this is the textual representation of a serialized PHP object, then yes that's would have been drastically easier. Thanks for pointing it out. I guess I'll never stop learning.

File: includes/polling/applications/memcached.inc.php [hunk 3]
```diff
@@ -1,6 +1,13 @@
 <?php
 
-$data = $agent_data['app']['memcached'][$app['app_instance']];
+if (!empty($agent_data['app']['memcached'])) {
+    $data = $agent_data['app']['memcached'][$app['app_instance']];
+} else {
+    $options = '-O qv';
+    $oid     = '.1.3.6.1.4.1.8072.1.3.2.3.1.2.9.109.101.109.99.97.99.104.101.100';
+    $result  = snmp_get($device, $oid, $options);
+    $data    = reset(unserialize(str_replace("<<<app-memcached>>>\n", '', $result)));
```
Comment (murrant) [inline includes/polling/applications/memcached.inc.php]: Scrutinizer bug, reset() expects a reference.
So you might as well set it to a variable and then use `[0]`

**Comment by murrant** [comment]
Looks good, just need to you to fix the merge conflicts likely the Contributer file.

**Pull request merged**